src/query/sql/src/planner/optimizer/rule/rewrite/rule_eliminate_subquery.rs

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use databend_common_exception::Result;
+use databend_common_expression::types::DataType;
+
+use crate::optimizer::extract::Matcher;
+use crate::optimizer::rule::Rule;
+use crate::optimizer::rule::TransformResult;
+use crate::optimizer::RuleID;
+use crate::optimizer::SExpr;
+use crate::plans::walk_expr_mut;
+use crate::plans::BoundColumnRef;
+use crate::plans::Exchange;
+use crate::plans::JoinType;
+use crate::plans::RelOperator;
+use crate::plans::VisitorMut;
+use crate::ColumnBinding;
+use crate::ColumnBindingBuilder;
+use crate::ColumnEntry;
+use crate::IndexType;
+use crate::MetadataRef;
+use crate::ScalarExpr;
+use crate::Visibility;
+
+pub struct RuleEliminateSubquery {
+    id: RuleID,
+    matchers: Vec<Matcher>,
+    pub(crate) metadata: MetadataRef,
+}
+
+impl RuleEliminateSubquery {
+    pub fn new(metadata: MetadataRef) -> Self {
+        Self {
+            id: RuleID::EliminateSubquery,
+            // Join (only for subquery)
+            // |  \
+            // *   *
+            matchers: vec![Matcher::MatchFn {
+                predicate: Box::new(|op| {
+                    if let RelOperator::Join(join) = op {
+                        return join.join_type == JoinType::LeftSemi
+                            && join.equi_conditions.len() == 1;
+                    }
+                    false
+                }),
+                children: vec![Matcher::Leaf, Matcher::Leaf],
+            }],
+            metadata,
+        }
+    }
+}
+
+impl Rule for RuleEliminateSubquery {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformResult) -> Result<()> {
```

Comment:
Better to add some unit tests for this rule apply, the logic test is not enough :/